staging/src/k8s.io/apiserver/pkg/storage/value/encrypt/aes/aes_cbc_hmac.go

```diff
@@ -0,0 +1,217 @@
+package aes
+
+import (
+	"bytes"
+	"crypto/cipher"
+	"crypto/hmac"
+	"crypto/rand"
+	"errors"
+	"hash"
+	"io"
+)
+
+// AEAD is an interface that mimics cipher.AEAD to implement an
+// Authenticated Encryption with Associated Data scheme.
+// This implements the AEAD scheme as defined in RFC5116 at:
+// https://tools.ietf.org/html/rfc5116
+type AEAD interface {
+	// NonceSize returns the size of the nonce that must be passed to Seal
+	// and Open.
+	NonceSize() int
+
+	// Overhead returns the maximum difference between the lengths of a
+	// plaintext and its ciphertext.
+	Overhead() int
+
+	// Seal encrypts and authenticates plaintext, authenticates the
+	// additional data and appends the result to dst, returning the updated
+	// slice. The nonce must be NonceSize() bytes long and unique for all
+	// time, for a given key.
+	//
+	// To reuse plaintext's storage for the encrypted output, use plaintext[:0]
+	// as dst. Otherwise, the remaining capacity of dst must not overlap plaintext.
+	Seal(dst, nonce, plaintext, additionalData []byte) []byte
+
+	// Open decrypts and authenticates ciphertext, authenticates the
+	// additional data and, if successful, appends the resulting plaintext
+	// to dst, returning the updated slice. The nonce must be NonceSize()
+	// bytes long and both it and the additional data must match the
+	// value passed to Seal.
+	//
+	// To reuse ciphertext's storage for the decrypted output, use ciphertext[:0]
+	// as dst. Otherwise, the remaining capacity of dst must not overlap plaintext.
+	//
+	// Even if the function fails, the contents of dst, up to its capacity,
+	// may be overwritten.
+	Open(dst, nonce, ciphertext, additionalData []byte) ([]byte, error)
+}
+
+// aesCbcHmac represents an AES block cipher in CBC mode authenticated with HMAC data
+type aesCbcHmac struct {
+	cipher    cipher.Block
+	hash      hash.Hash
+	nonceSize int
+	tagSize   int
+}
+
+const (
+	achBlockSize         = 16
+	achTagSize           = 16
+	achMinimumTagSize    = 12 // NIST SP 800-38D recommends tags with 12 or more bytes.
+	achStandardNonceSize = 12
+)
+
+// NewACH returns and AEAD which uses the AES block cipher in CBC mode authenticated with HMAC data
+func NewACH(cipher cipher.Block, hash hash.Hash) (AEAD, error) {
+	return newACHWithNonceAndTagSize(cipher, hash, achStandardNonceSize, achTagSize)
+}
+
+func newACHWithNonceAndTagSize(cipher cipher.Block, hash hash.Hash, nonceSize int, tagSize int) (AEAD, error) {
+	if tagSize < achMinimumTagSize || tagSize > achBlockSize {
```

Comment:
is a check needed for tagSize <= hash.Size()? can //Truncate HMAC to tagsize result in  bounds out of range if not?